Add support of local environment variables in kustomize lookup plugin

yurnov opened the pull request:
##### SUMMARY
kustomize doesn't support an environment that makes it impossible to use `HTTP_PROXY` or provide some templatized parameters. 

This PR is the result of the issue #783 

##### ISSUE TYPE
- Feature Pull Request


##### COMPONENT NAME
kubernetes.core.kustomize lookup plugin

yurnov commented:
Hi @abikouo,

could you please check for workflow, some sanity tests are failing on `milestone` and `development` and it seems that it is unrelated to this PR reason.

yurnov commented:
> Hi @abikouo,
> 
> could you please check for workflow, some sanity tests are failing on `milestone` and `development` and it seems that it is unrelated to this PR reason.

I found that the root cause of the sanity failure was switching the `devel` and `milestone` branches of Ansible to version 2.19. The test was fixed by commit deff6eeec4196ad7b81c7f098e41ad4c025b12b3, which was included in that PR. But I think I should probably create a new PR for CI (GitHub actions) related issues to have a much cleaner history.

yurnov commented:
PR #789 was created as I mentioned in [previous comment](https://github.com/ansible-collections/kubernetes.core/pull/786#issuecomment-2452996442), this PR should be rebased when 789 is merged.

abikouo commented on plugins/lookup/kustomize.py:
```diff
@@ -34,6 +34,12 @@
         description:
         - Enable the helm chart inflation generator
         default: "False"
+      enviroment:
```
```suggestion
      environment:
```

yurnov replied:
Done

abikouo commented on plugins/lookup/kustomize.py:
```diff
@@ -55,6 +61,10 @@
+
+- name: Create kubernetes resources for lookup output with environment variables
+  kubernetes.core.k8s:
+    definition: "{{ lookup('kubernetes.core.kustomize', binary_path='/path/to/kubectl', enviroment='HTTP_PROXY=http://proxy.example.com:3128') }}"
```
```suggestion
    definition: "{{ lookup('kubernetes.core.kustomize', binary_path='/path/to/kubectl', environment='HTTP_PROXY=http://proxy.example.com:3128') }}"
```

yurnov replied:
Done

yurnov replied:
Done

abikouo commented on plugins/lookup/kustomize.py:
```diff
@@ -141,7 +155,22 @@ def run(
+                    key, value = env.split("=")
+                    environ[key] = value
+            if isinstance(enviroment, dict):
+                for key, value in enviroment.items():
+                    environ[key] = value
```
```suggestion
                environ.update(environment)
```

yurnov replied:
Done

yurnov replied:
Done

abikouo commented on plugins/lookup/kustomize.py:
```diff
@@ -141,7 +155,22 @@ def run(
             command += ["--enable-helm"]
 
-        (ret, out, err) = run_command(command)
+        environ = os.environ.copy()
+
+        if enviroment:
```
```suggestion
        environ = None

        if enviroment:
           environ = os.environ.copy()
```

yurnov replied:
Done

yurnov replied:
Done

abikouo reviewed (changes requested):
@yurnov thanks for working on this issue.
Some minor things need to be changed.
If possible add some integration tests case for this feature into `lookup_kustomize` integration test target

yurnov commented:
I will check what can be done to integration test

yurnov commented:
Hi @abikouo,

I have fixed the typo, considered your suggestion, and added integration tests with positive and negative scenarios. Please review PR again.

For unclear reasons, kubevirt test is failing.

abikouo commented on plugins/lookup/kustomize.py:
```diff
@@ -34,6 +34,12 @@
+        - The environment variables to pass to the kustomize or kubectl command.
+        type: dict
+        default: {}
+        version_added: 6.0.0
```
```suggestion
        version_added: 6.2.0
```

abikouo commented on plugins/lookup/kustomize.py:
```diff
@@ -55,6 +61,10 @@
   kubernetes.core.k8s:
     definition: "{{ lookup('kubernetes.core.kustomize', dir='/path/to/kustomization', enable_helm=True) }}"
+
+- name: Create kubernetes resources for lookup output with environment variables
```
Could you please add an example with a multiple-key dictionary? The attribute is supposed to be of type `dict`, but here you are defining it like `key=value`

yurnov replied:
done

abikouo commented on plugins/lookup/kustomize.py:
```diff
@@ -141,7 +155,21 @@ def run(
+        environ = None
+        if environment:
+            environ = os.environ.copy()
+            if isinstance(environment, str):
```
The code here assumes that the value should be defined as `raw`, which can either be a `dict` or a `str` with a specific syntax

yurnov replied:
updated

abikouo commented on plugins/lookup/kustomize.py:
```diff
@@ -141,7 +155,21 @@ def run(
+                    raise AnsibleLookupError(
+                        "environment should be dict or string in the format key=value"
+                    )
+                for env in environment.split(" "):
```
Please add to the definition that a space should separate multiple key/value elements

yurnov replied:
done

abikouo replied:
the plugin documentation has not been updated
